src/main/java/randoop/generation/Bloodhound.java

```diff
@@ -0,0 +1,237 @@
+   * method has been selected.
+   *
+   * @return the chosen {@code TypedOperation} for the new sequence
+   */
+  @Override
+  public TypedOperation selectOperation() {
+    // Collect branch coverage and recompute weights for methods under test.
+    updateBranchCoverageMaybe();
+
+    // Make a random, weighted choice for the next method
+    TypedOperation operation = Randomness.randomMemberWeighted(operationSimpleList, methodWeights);
+
+    // Update its selection count and its weight.
+    incrementInMap(methodSelectionCounts, operation);
+    updateWeightForOperation(operation);
+
+    return operation;
+  }
+
+  /**
+   * The branch coverage information for all methods under test is re-summarized at every {@code
+   * branchCoverageInterval}'th call of this method. Weights for all methods under test are
+   * recomputed when branch information is re-summarized.
+   */
+  private void updateBranchCoverageMaybe() {
+    stepNum += 1;
+    if (stepNum % branchCoverageInterval == 0) {
+      methodSelectionCounts.clear();
+      CoverageTracker.instance.summarizeCoverageInformation();
+      updateWeightsForAllOperations();
+    }
+  }
+
+  /** Computes and updates weights in our method weights map for all methods under test. */
+  private void updateWeightsForAllOperations() {
+    for (TypedOperation operation : operationSimpleList) {
+      updateWeightForOperation(operation);
+    }
+  }
+
+  /**
+   * Recompute weights for a method under test. A method under test is assigned a weight based on a
+   * weighted combination of the number of branches uncovered and the ratio between the number of
+   * times this method has been selected and the maximum number of times any method under test has
+   * been selected. The weighting scheme is based on Bloodhound in the Guided Random Testing (GRT)
+   * paper.
+   *
+   * @param operation method to compute weight for
+   */
+  private void updateWeightForOperation(TypedOperation operation) {
+    // The number of methods under test, corresponds to |M| in the GRT paper.
+    int numOperations = this.operationSimpleList.size();
+
+    // Default weight is uniform probability.  It is used if no coverage details are available.
+    // This is the case for classes like java.lang.Object which is not explicitly under test.
+    double weight = 1.0 / numOperations;
+
+    CoverageTracker.BranchCoverage covDet =
+        CoverageTracker.instance.getDetailsForMethod(operation.getName());
+
+    // Check that branch coverage details are available for this method.
+    if (covDet != null) {
+      // The number of successful invocations of this method. Corresponds to succ(m).
+      Integer numSuccessfulInvocations = methodInvocationCounts.get(operation);
+      if (numSuccessfulInvocations == null) {
+        numSuccessfulInvocations = 0;
+      }
+
+      // If this method has no uncovered branches, or it is the most-selected method,
```

Comment:
Why?  The paper does not have a special case for the most-selected method, so this seems to differ from the paper's formula.
I suggest that you make this method as simple as possible, and make it obviously equivalent to the formula in the paper.

Reply:
Hi Mike,
This extra clause is meant to catch the case where `uncovRatio(m) == 0` and `succ(m) / maxSucc(M) == 1` in which case `w(m, 0)` would evaluate to zero.
![image](https://user-images.githubusercontent.com/15660492/38847567-2d8067f8-41b7-11e8-8e95-f95bea29c4eb.png)
This is problematic because `Randomness.randomMemberWeighted` in Randoop does not allow an element's weight to be zero. I am thinking that it might make sense to treat this as an uncovered method and give it a default weight, or adjust random member selection to ignore elements with zero weight. Should I further document this choice or make the change?